src/sys/wasi/mod.rs

```diff
@@ -0,0 +1,363 @@
+            stream.set_nonblocking(true)?;
+            Ok((stream, addr))
+        }
+    }
+}
+
+/// Unique id for use as `SelectorId`.
+#[cfg(debug_assertions)]
+#[cfg(feature = "net")]
+static NEXT_ID: AtomicUsize = AtomicUsize::new(1);
+
+pub(crate) struct Selector {
+    #[cfg(debug_assertions)]
+    #[cfg(feature = "net")]
+    id: usize,
+    /// Subscriptions (reads events) we're interested in.
+    subscriptions: Arc<Mutex<Vec<wasi::Subscription>>>,
+}
+
+impl Selector {
+    pub(crate) fn new() -> io::Result<Selector> {
+        Ok(Selector {
+            #[cfg(debug_assertions)]
+            #[cfg(feature = "net")]
+            id: NEXT_ID.fetch_add(1, Ordering::Relaxed),
+            subscriptions: Arc::new(Mutex::new(Vec::new())),
+        })
+    }
+
+    #[cfg(debug_assertions)]
+    #[cfg(feature = "net")]
+    pub(crate) fn id(&self) -> usize {
+        self.id
+    }
+
+    pub(crate) fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        events.clear();
+
+        let mut subscriptions = self.subscriptions.lock().unwrap();
+
+        // If we want to a use a timeout in the `wasi_poll_oneoff()` function
+        // we need another subscription to the list.
+        if let Some(timeout) = timeout {
+            subscriptions.push(timeout_subscription(timeout));
+        }
+
+        // `poll_oneoff` needs the same number of events as subscriptions.
+        let length = subscriptions.len();
+        events.reserve(length);
+
+        debug_assert!(events.capacity() >= length);
+
+        let res = unsafe { wasi::poll_oneoff(subscriptions.as_ptr(), events.as_mut_ptr(), length) };
+
+        // Remove the timeout subscription we possibly added above.
+        if timeout.is_some() {
+            let timeout_sub = subscriptions.pop();
+            debug_assert_eq!(
+                timeout_sub.unwrap().u.tag,
+                wasi::EVENTTYPE_CLOCK.raw(),
+                "failed to remove timeout subscription"
+            );
+        }
+
+        drop(subscriptions); // Unlock.
+
+        match res {
+            Ok(n_events) => {
+                // Safety: `poll_oneoff` initialises the `events` for us.
+                unsafe { events.set_len(n_events) };
+
+                // Remove the timeout event.
+                if timeout.is_some() {
+                    if let Some(index) = events.iter().position(is_timeout_event) {
+                        events.swap_remove(index);
+                    }
+                }
+
+                check_errors(&events)
+            }
+            Err(err) => Err(io_err(err)),
+        }
+    }
+
+    #[cfg(feature = "net")]
+    pub(crate) fn register(
+        &self,
+        fd: wasi::Fd,
+        token: Token,
+        interests: Interest,
+    ) -> io::Result<()> {
+        let mut subscriptions = self.subscriptions.lock().unwrap();
+
+        // Cannot subscribe to read and write at the same time.
```

Comment:
This is still a blocking problem. It breaks Mio's design.

Reply:
That should be brought up upstream then, because that's how it currently works or better does not work.

Reply:
@haraldh What is upstream in this case? Is it `wasmtime` or `WASI`?

Reply:
It is possible to fix `wasmtime`. A fix for `WASI` is highly unlikely.

Reply:
Perhaps you could open an issue here https://github.com/WebAssembly/WASI.

Reply:
https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/witx/typenames.witx#L486-L499

Reply:
maybe push 2 subscriptions then?

Reply:
@Thomasdezeeuw There will be no new changes WASI snapshots until modularization is complete. So we have to work with what we have. Please note that I'm the person who pushed `sock_accept()` through the committee.

Reply:
> maybe push 2 subscriptions then?

If I remember correctly that doesn't work, but you can try.

> @Thomasdezeeuw There will be no new changes WASI snapshots until modularization is complete. So we have to work with what we have.

Is their an ETA for that? Because if we can't call `register` with both read and write interest Mio simply won't work with WASI.

> Please note that I'm the person who pushed `sock_accept()` through the committee.

Good to know 👍

Reply:
To be clear if creating 2 subscriptions work we're all good, kqueue also needs separate events: https://github.com/tokio-rs/mio/blob/e8aed7baea88e844a7402a29ec8675ddc341edd9/src/sys/unix/selector/kqueue.rs#L133-L147.

Reply:
@Thomasdezeeuw Unfortunately, there is no ETA for modularization.

Fortunately, I don't see anything in the WASI spec that makes multiple subscriptions invalid. So I propose we treat this as a `wasmtime` bug, if the bug exists at all. I suggest we make the minor change to allow simultaneous read and write subscriptions and if `wasmtime` chokes on this we file a followup PR to `wasmtime` under the assumption they are non-compliant.

Reply:
@haraldh Any update?

Reply:
It is wasmtime... see https://github.com/bytecodealliance/wasmtime/pull/3866

Reply:
I added one commit, which made polling on read and write possible with a patched wasmtime.

Reply:
wasmtime PR was merged 😀